crypto/slip39.c

```diff
@@ -65,34 +63,64 @@ bool word_index(uint16_t* index, const char* word, uint8_t word_length) {
 }
 
 /**
- * Calculates which buttons on the T9 keyboard can still be pressed after the
- * prefix was entered. Returns a 9-bit bitmask, where each bit specifies which
- * buttons can be pressed (there are still words in this combination). The least
- * significant bit corresponds to the first button.
- *
- * Example: 110000110 - second, third, eighth and ninth button still can be
- * pressed.
+ * Returns the index of the first sequence in words_button_seq[] which is not
+ * less than the given sequence. Returns WORDS_COUNT if there is no such
+ * sequence.
  */
-uint16_t slip39_word_completion_mask(uint16_t prefix) {
-  return find(prefix, false);
+static uint16_t find_sequence(uint16_t sequence) {
+  if (sequence <= words_button_seq[0].sequence) {
+    return 0;
+  }
+
+  uint16_t lo = 0;
+  uint16_t hi = WORDS_COUNT;
+
+  while (hi - lo > 1) {
+    uint16_t mid = (hi + lo) / 2;
+    if (words_button_seq[mid].sequence >= sequence) {
+      hi = mid;
+    } else {
+      lo = mid;
+    }
+  }
+
+  return hi;
 }
 
 /**
  * Returns the first word matching the button sequence prefix or NULL if no
  * match is found.
  */
-const char* button_sequence_to_word(uint16_t prefix) {
-  return get_word(find(prefix, true));
+const char* button_sequence_to_word(uint16_t sequence) {
```

Comment:
First, the definition of the function changed. The function used to return the first word in alphabetical order matching `sequence`, now it returns the word with the lowest prefix that matches the `sequence`. I guess it will change the word that is suggested to the user.

Second, the function returns a nonsense when the prefix contains the digit zero (such as `1011`). It should be ok, since the function is never called with such a prefix. Maybe it should be at least in comments.

Reply:
> First, the definition of the function changed. The function used to return the first word in alphabetical order matching `sequence`, now it returns the word with the lowest prefix that matches the `sequence`. I guess it will change the word that is suggested to the user.

That's a good catch!

Reply:
> First, the definition of the function changed. The function used to return the first word in alphabetical order matching `sequence`, now it returns the word with the lowest prefix that matches the `sequence`. I guess it will change the word that is suggested to the user.

Trezor only shows the word once there is a unique match. So this makes no difference as far as I could tell.

> Ideally we could create a unit-test for this function.

There is a unit test `test_slip39_sequence_to_word`, which I originally updated in 7ce5a2b precisely because of this.

> Second, the function returns a nonsense when the prefix contains the digit zero (such as `1011`). It should be ok, since the function is never called with such a prefix. Maybe it should be at least in comments.

Good point, but this is only a symptom of a more general problem, e.g. sequence = 255, although that is also unreachable at the moment. Nevertheless fixed in 71975ea and added more unit tests in 2c65bc4.

Reply:
f69800e: Also added the same unit tests to `test_slip39_word_completion_mask`.

Reply:
> Trezor only shows the word once there is a unique match. So this makes no difference as far as I could tell.

You are right, I was confused, because it works differently for slip39 backup and bip39 backup.